XtermSharp/InputHandlers/InputHandler.cs

```diff
@@ -1200,56 +1200,75 @@ unsafe void Print (byte* data, int start, int end)
 			var curAttr = terminal.CurAttr;
 			var bufferRow = buffer.Lines [buffer.Y + buffer.YBase];
 
-
 			terminal.UpdateRange (buffer.Y);
 
-			while (readingBuffer.HasNext ()) {
-				int code;
-				byte bufferValue = readingBuffer.GetNext ();
-				var n = RuneExt.ExpectedSizeFromFirstByte (bufferValue);
-				if (n == -1) {
-					// Invalid UTF-8 sequence, client sent us some junk, happens if we run with the wrong locale set
-					// for example if LANG=en
-					code = (int)((uint)bufferValue);
-				} else if (n == 1) {
-					code = bufferValue;
-				} else {
-					var bytesRemaining = readingBuffer.BytesLeft ();
-					if (readingBuffer.BytesLeft () >= (n - 1)) {
-						var x = new byte [n];
-						x [0] = bufferValue;
-						for (int j = 1; j < n; j++)
-							x [j] = readingBuffer.GetNext ();
-
-						(var r, var size) = Rune.DecodeRune (x);
-						code = (int)(uint)r;
-					} else {
-						readingBuffer.Putback (bufferValue);
-						return;
+			while (readingBuffer.HasNext()) {
+				var ch = ' ';
+				var chWidth = 0;
+				int code = 32;
+				byte bufferValue = readingBuffer.GetNext();
+				var n = RuneExt.ExpectedSizeFromFirstByte(bufferValue);
+				if (n == -1 || n == 1) {
+					var chSet = false;
+					if (bufferValue < 127 && charset != null) {
+						var str = charset[bufferValue];
+						if (!string.IsNullOrEmpty(str)) {
+							ch = str[0];
+							// Every single mapping in the charset only takes one slot
+							chWidth = 1;
+							chSet = true;
+						}
 					}
-				}
 
-				// MIGUEL-TODO: I suspect this needs to be a stirng in C# to cope with Grapheme clusters
-				var ch = code;
+					if (!chSet) {
+						(var rune, var size) = Rune.DecodeRune(new byte[] { bufferValue });
```

Comment:
Another candidate for reuse, we can allocate the byte array above the loop here.